### DISCUSSION TITLE
Need help with code in blog post

### DISCUSSION THREAD

**Discussion by jsolderitsch** [deepgram/community#98] [General help] [answered]
I am following along with the code in this blog post:
[https://blog.deepgram.com/how-to-transcribe-only-what-you-need-with-python-listening-before-connected/](https://blog.deepgram.com/how-to-transcribe-only-what-you-need-with-python-listening-before-connected/)
The code causes the following error:
```
Error while sending: Task <Task pending name='Task-8' coro=<process.<locals>.sender() running at /Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py:50> cb=[gather.<locals>._done_callback() at /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py:764]> got Future <Future pending> attached to a different loop
Traceback (most recent call last):
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py", line 72, in <module>
    asyncio.run(run())
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py", line 69, in run
    await asyncio.gather(microphone(),process())
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py", line 64, in process
    await asyncio.gather(sender(ws), receiver(ws))
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py", line 50, in sender
    data = await audio_queue.get()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/queues.py", line 166, in get
    await getter
RuntimeError: Task <Task pending name='Task-8' coro=<process.<locals>.sender() running at /Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/stream0.py:50> cb=[gather.<locals>._done_callback() at /Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py:764]> got Future <Future pending> attached to a different loop
```
Running on a Mac with the Monterey OS and Python 3.9.5. I named the code file stream0.py

I did not change any of the organization in the code sample but added my own Deepgram token string.
Perhaps someone at Deepgram can offer a suggestion?

**Comment by briancbarrow** [comment]
Hi @jsolderitsch did you run the program by itself or as part of a larger program? It looks like there is a typo on line 51. The trailing `.` shouldn't be there. Other than that, I ran the code in a new project and it worked. 

Can you provide more context around how you are running the code?

**Reply by jsolderitsch** [reply]
I did correct the error -- removed the trailing `.`

Running as is -- not part of a larger program.

What PC are you using to run it? What version of Python?

I am using Visual Studio Code on  a Mac running Monterey OS version as I said in my post. What do you mean by "New Project"?

In the VSC terminal I just execute: `python stream0.py`  And there I get the error I reported.

What am I doing wrong?

**Reply by jsolderitsch** [reply]
Actually -- can you tell me what happens to you when you run successfully? What should I be seeing in the terminal where I run python when it doesn't cause an exception trace?

**Reply by briancbarrow** [reply]
I'm on a Mac running Monterey OS with python version 3.10.7

I am using Visual Studio Code also and ran it the same way you did. When you run it, you should get a prompt asking for access to your microphone and then the terminal will display the transcript it captures from your microphone. Here's what mine looked like. (I named my file `blog.py`)
<img width="577" alt="Screen Shot 2023-03-15 at 2 03 37 PM" src="https://user-images.githubusercontent.com/13695093/225429369-5be59fec-f011-4baf-92c4-37db2a4fe08c.png">


Did you get all the code?

```python
import pyaudio
import asyncio
import websockets
import os
import json

DEEPGRAM_API_KEY = "YOUR_DEEPGRAM_API_KEY"

FORMAT = pyaudio.paInt16
CHANNELS = 1
RATE = 16000
CHUNK = 8000

audio_queue = asyncio.Queue()

def callback(input_data, frame_count, time_info, status_flags):
   audio_queue.put_nowait(input_data)

   return (input_data, pyaudio.paContinue)


async def microphone(): 
   audio = pyaudio.PyAudio()
   stream = audio.open(
       format = FORMAT,
       channels = CHANNELS,
       rate = RATE,
       input = True,
       frames_per_buffer = CHUNK,
       stream_callback = callback
   )

   stream.start_stream()

   while stream.is_active():
       await asyncio.sleep(0.1)


   stream.stop_stream()
   stream.close()

async def process():
   extra_headers = {
       'Authorization': 'token ' + DEEPGRAM_API_KEY
   }

   async with websockets.connect('wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=16000&channels=1', extra_headers = extra_headers) as ws:
       async def sender(ws): # sends audio to websocket
           try:
               while True:
                   data = await audio_queue.get()
                   await ws.send(data)
           except Exception as e:
               print('Error while sending: ', + str(e))
               raise

       async def receiver(ws): 
           async for msg in ws:
               msg = json.loads(msg)
               transcript = msg['channel']['alternatives'][0]['transcript']

               if transcript:
                   print(f'Transcript = {transcript}')

       await asyncio.gather(sender(ws), receiver(ws))

      

async def run():
   await asyncio.gather(microphone(),process())

if __name__ == '__main__':
   asyncio.run(run())
```

**Reply by jsolderitsch** [reply]
OK, different error now:
```
Traceback (most recent call last):
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/blog.py", line 73, in <module>
    asyncio.run(run())
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/blog.py", line 70, in run
    await asyncio.gather(microphone(),process())
  File "/Users/jjs/Controller-DroneBlocks/VoiceControlledCar/Code/Python/blog.py", line 47, in process
    async with websockets.connect('wss://api.deepgram.com/v1/listen?encoding=linear16&sample_rate=16000&channels=1', extra_headers = extra_headers) as ws:
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/websockets/legacy/client.py", line 642, in __aenter__
    return await self
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/websockets/legacy/client.py", line 659, in __await_impl_timeout__
    return await asyncio.wait_for(self.__await_impl__(), self.open_timeout)
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/asyncio/tasks.py", line 481, in wait_for
    return fut.result()
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/websockets/legacy/client.py", line 666, in __await_impl__
    await protocol.handshake(
  File "/Library/Frameworks/Python.framework/Versions/3.9/lib/python3.9/site-packages/websockets/legacy/client.py", line 332, in handshake
    raise InvalidStatusCode(status_code, response_headers)
websockets.exceptions.InvalidStatusCode: server rejected WebSocket connection: HTTP 401
```
???

**Reply by jsolderitsch** [reply]
I am on campus now so a different wireless config than my home network. IT has things more locked down here?

**Reply by briancbarrow** [reply]
401 is just an anauthorized error. Do you still have the correct API key in your code?

**Comment by jsolderitsch** [comment, marked as answer]
And you are running a newer Python!

**Reply by briancbarrow** [reply]
I'm running python 3.10.7

**Comment by jsolderitsch** [comment]
I am using the API "secret" as the key -- is that the right thing to use?

Jim

On Mar 15, 2023, at 4:31 PM, Brian Barrow ***@***.******@***.***>> wrote:



401 is just an anauthorized error. Do you still have the correct API key in your code?

—

**Reply by briancbarrow** [reply]
Yeah the secret is the right thing. Can you try creating a new API key and using it instead?

**Reply by jsolderitsch** [reply]
Did not need to do that. The problem was fixed after installing python 3.10.10 (the latest 3.10 version available) and then using this line
```
async def process():
   extra_headers = {
       'Authorization': 'token ' + DEEPGRAM_API_KEY
   }
```
I had somehow quoted the DEEPGRAM_API_KEY before when I pulled in your code.

When I fixed this the HTTP 401 error went away but then the other exception trace behavior returned.

So then I got a python 3.10 version and then the code works.

I recommend the blog post text be edited to warn about python executable versions.

Thanks for your help.

**Reply by briancbarrow** [reply]
Thanks @jsolderitsch. Glad you got it working. I'll make a note to get that post updated.

**Reply by jsolderitsch** [reply]
Thanks for your help. I wonder if other platforms like Windows or even the Raspberry Pi have sensitivity to the version of Python. I see that version 3.11.2 of Python is released now. But many platforms have OS-related scripts that are Python version dependent so they can move along as rapidly when new Python versions get released. I know that Ubuntu and the Raspberry Pi OS are like this. Would I have to go through the trouble of building a newer Python to run the blog code there or is the base Python more forgiving of all of the async calls and organization present in the blog code using web sockets.
